### PR TITLE
Optimisations 3.13.1

### PR DESCRIPTION
**Pull request by gudnimg**
I went through all the commits in https://github.com/prusa3d/Prusa-Firmware/pull/3378/commits and pulled in the changes which are still relevant. IMO I think creating a new PR would be less work than trying to rebase the original PR as we have made soo many changes.

Change in memory:
Flash: -168 bytes
SRAM: 0 bytes

Additionally 148 bytes of Flash memory are saved when `TMC2130_SERVICE_CODES_M910_M918` is enabled

### CODE REVIEW

File: Firmware/lcd.cpp [hunk 1]
```diff
@@ -129,18 +129,17 @@ static void lcd_send(uint8_t data, uint8_t flags, uint16_t duration = LCD_DEFAUL
 	_delay_us(5);
 	lcd_writebits(data);
 #ifndef LCD_8BIT
-	if (!(flags & LCD_HALF_FLAG))
-	{
-		_delay_us(LCD_DEFAULT_DELAY);
-		lcd_writebits(data<<4);
+	if (!(flags & LCD_HALF_FLAG)) {
+		// _delay_us(LCD_DEFAULT_DELAY); // should not be needed when sending a two nibble instruction.
+		lcd_writebits((data << 4) | (data >> 4)); //force efficient swap opcode even though the lower nibble is ignored in this case
```
Comment (DRracer) [inline Firmware/lcd.cpp:134]: :+1: interesting

**Review by DRracer — APPROVED** [review]

File: Firmware/tmc2130.cpp [hunk 2]
```diff
@@ -867,13 +867,10 @@ void tmc2130_get_wave(uint8_t axis, uint8_t* data, FILE* stream)
...
-		{
-			if (mscnt == i)
-				fprintf_P(stream, PSTR("%d\t%d\n"), i, curA);
-			else //TODO - remove this check
-				fprintf_P(stream, PSTR("!! (i=%d MSCNT=%d)\n"), i, mscnt);
-		}
+		if (mscnt == i)
+			printf_P(PSTR("%d\t%d\n"), i, curA);
+		else //TODO - remove this check
+			printf_P(PSTR("!! (i=%d MSCNT=%d)\n"), i, mscnt);
```
Comment (3d-gussner) [inline Firmware/tmc2130.cpp]: Please change `!!` to single `!` to prevent OctoPrint to ERROR out

File: Firmware/Dcodes.cpp
Comment (3d-gussner) [inline Firmware/Dcodes.cpp:1]: Add `#include "util.h"` to fix build with `#define DEBUG_CODES` active

**Review by 3d-gussner — CHANGES_REQUESTED** [review]
Tested
- TMC changes with `D2130` :heavy_check_mark:
- Language changes :heavy_check_mark: 
- Crash menu :heavy_check_mark:

File: Firmware/tmc2130.cpp (hunk 2 above)
Comment (gudnimg) [reply Firmware/tmc2130.cpp]: Done

File: Firmware/Dcodes.cpp
Comment (gudnimg) [reply Firmware/Dcodes.cpp:1]: Done

**Review by 3d-gussner — APPROVED** [review]
Only thing to test bit more is the LCD timing change on real hardware. The rest looks good now.

**Pull request merged**